src/Nethermind/Nethermind.TxPool/TxPool.cs

```diff
@@ -106,31 +105,40 @@ public class TxPool : ITxPool, IDisposable
         /// (by miners or validators) or simply informing other nodes about known pending transactions (broadcasting).
         /// </summary>
         /// <param name="txStorage">Tx storage used to reject known transactions.</param>
-        /// <param name="timestamper">Used for calculating the difference between the current time and the time when the transaction was added.</param>
         /// <param name="ecdsa">Used to recover sender addresses from transaction signatures.</param>
         /// <param name="specProvider">Used for retrieving information on EIPs that may affect tx signature scheme.</param>
         /// <param name="txPoolConfig"></param>
         /// <param name="stateProvider"></param>
         /// <param name="logManager"></param>
-        public TxPool(
-            ITxStorage txStorage,
-            ITimestamper timestamper,
+        /// <param name="comparer"></param>
+        public TxPool(ITxStorage txStorage,
             IEthereumEcdsa ecdsa,
             ISpecProvider specProvider,
             ITxPoolConfig txPoolConfig,
             IStateProvider stateProvider,
-            ILogManager logManager)
+            ILogManager logManager,
+            IComparer<Transaction> comparer = null)
         {
             _ecdsa = ecdsa ?? throw new ArgumentNullException(nameof(ecdsa));
             _logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
             _txStorage = txStorage ?? throw new ArgumentNullException(nameof(txStorage));
-            _timestamper = timestamper ?? throw new ArgumentNullException(nameof(timestamper));
             _specProvider = specProvider ?? throw new ArgumentNullException(nameof(specProvider));
             _stateProvider = stateProvider ?? throw new ArgumentNullException(nameof(stateProvider));
 
             MemoryAllowance.MemPoolSize = txPoolConfig.Size;
             ThisNodeInfo.AddInfo("Mem est tx   :", $"{(LruCache<Keccak, object>.CalculateMemorySize(32, MemoryAllowance.TxHashCacheSize) + LruCache<Keccak, Transaction>.CalculateMemorySize(4096, MemoryAllowance.MemPoolSize)) / 1000 / 1000}MB".PadLeft(8));
-            _transactions = new DistinctValueSortedPool<Keccak, Transaction>(MemoryAllowance.MemPoolSize, (t1, t2) => t1.GasPrice.CompareTo(t2.GasPrice), PendingTransactionComparer.Default);
+            
+            // we need to ensure transactions are ordered by nonce, which might not be done in supplied comparer
+            comparer = new NonceTransactionComparerDecorator(comparer ?? GasBasedTxComparer.Instance);
+            _transactions = new DistinctValueSortedPool<Keccak, Transaction, Address>(
+                MemoryAllowance.MemPoolSize, 
+                // in order to sort properly and not loose transactions we need to differentiate on their identity which provided comparer might not be doing
+                new TxIdentityCompositeDecorator(comparer),
```

Comment:
comparer.Then(ByHashComparer.Instance) => it would create a CompositeComparer(comparer, ByHashComparer)

Reply:
so you can use an extension method that improves readability instead of nesting without clear comparison ordering

Reply:
abstracted it